### PR TITLE
feat: force kid with public prefix

### PR DESCRIPTION
**Pull request by naveenpaul1**
## Related issue

https://github.com/ory/hydra/discussions/2435 
@aeneasr  @Benehiko 

## Proposed changes

remove prefix public from key which is exposed in  .well-known/jwks.json

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [] I have added tests that prove my fix is effective or that my feature
      works.
- [] I have added or changed [the documentation](docs/docs).

@aeneasr

### CODE REVIEW

**Comment by aeneasr** [comment]
I think this could be breaking some stuff - I have to look into this but currently am not able due to time constraints...

File: jwk/helper.go
Comment (aeneasr) [inline jwk/helper.go:144]: We can actually use type assertion here on jose.JSONWebKey to check if the thing is a private key or not (like here https://github.com/ory/hydra/pull/2459/files#diff-d2d06a70d7e57e038f22beffe343a39d7efa66e3be691e00342697c5b7a909c9R125 ). I think that would be even better than doing this prefix stuff here.

**Review by aeneasr — CHANGES_REQUESTED** [review]
Sorry for taking ages on reviewing this. There's a couple of things we should address:

1. We could use type assertions to filter public keys instead
2. Need to add tests

File: jwk/helper.go
Comment (naveenpaul1) [reply jwk/helper.go:144]: @aeneasr : will do the changes

**Comment by naveenpaul1** [comment]
@aeneasr : Code changes are done, could you please review the same

**Review by aeneasr — APPROVED** [review]

**Comment by aeneasr** [comment]
Yup, waiting for CI to pass!

File: jwk/helper.go
Comment (aeneasr) [inline jwk/helper.go:148]: Could you please add a test for these two functions which:

1. Generate an RSA private/public keypair
2. Generate an ECDSA private/public keypair
3. Run `FindPublicKey` and `ExcludePrivateKeys`
4. Expect no private keys to be exposed, but public keys to be exposed

File: jwk/handler_test.go
Comment (aeneasr) [inline jwk/handler_test.go:90]: Shouldn't this be `1`? The first key is a private key

File: jwk/helper.go
Comment (aeneasr) [inline jwk/helper.go:142]: This only works for RSA keys, but what if we use ECDSA? This should not exclude private keys, it should only include "allow listed" public keys.

**Review by aeneasr — CHANGES_REQUESTED** [review]
Awesome, thank you for your contribution! This looks pretty good and I have some ideas how to improve it further :)

File: jwk/helper.go
Comment (naveenpaul1) [reply jwk/helper.go:142]: Let me check that @aeneasr

**Comment by aeneasr** [comment]
While the PR is being worked on I will mark it as a draft. That declutters our review backlog :)

Once you're done with your changes and would like someone to review them, mark the PR as ready and request a review from one of the maintainers.

Thank you!

File: jwk/helper_test.go
Comment (TimoDoernenburg) [inline jwk/helper_test.go:75]: created a new test case to find public RSA and ECDSA keys.

File: jwk/helper.go
Comment (TimoDoernenburg) [inline jwk/helper.go]: Added ecdsa keys to the loop

File: jwk/helper.go
Comment (aeneasr) [inline jwk/helper.go:147]: This list should not filter private keys out, it should only keep public keys in :)

**Review by aeneasr — CHANGES_REQUESTED** [review]

File: jwk/helper.go
Comment (TimoDoernenburg) [reply jwk/helper.go:147]: It's a bit unclear for me what I should change in this function. 
Should the function delete the private keys from the existing list instead of copying the public keys into another list?

Comment (TimoDoernenburg) [reply jwk/helper.go:147]: Something like this ? 

`func ExcludePrivateKeys(set *jose.JSONWebKeySet) *jose.JSONWebKeySet {
	keys := new(jose.JSONWebKeySet)
	i := 0 // output index
	for _, k := range set.Keys {
		_, ecdsaOk := k.Key.(*ecdsa.PublicKey)
		_, dsaOK := k.Key.(*dsa.PublicKey)
		_, ed25519OK := k.Key.(*ed25519.PublicKey)
		_, rsaOK := k.Key.(*rsa.PublicKey)

		if !(ecdsaOk || dsaOK || ed25519OK || rsaOK) {
			set.Keys[i] = k
			i++
		}
	}
	return keys
}`

Comment (aeneasr) [reply jwk/helper.go:147]: Exactly! Sorry for the late reply.

File: jwk/helper.go
Comment (TimoDoernenburg) [inline jwk/helper.go:141]: Updated the function based on the last comment.

**Comment by aeneasr** [comment]
Thank you, however it looks like test are not passing. Could you please address the failures?

File: jwk/helper_test.go
Comment (aeneasr) [inline jwk/helper_test.go:84]: One last thing, please ensure here that the key is indeed a public key:


```suggestion
		assert.IsType(t, keys.Key, new(rsa.PublicKey))
```

File: jwk/helper_test.go
Comment (aeneasr) [inline jwk/helper_test.go:91]:
```suggestion
		assert.IsType(t, keys.Key, new(ecdsa.PublicKey))
```

**Review by aeneasr — CHANGES_REQUESTED** [review]

**Comment by TimoDoernenburg** [comment]
The tests are passing now.

**Pull request closed without merging**
